coverage/docs/design.md

```diff
@@ -0,0 +1,165 @@
+ in artifacts directory.
+    - Completion marker is used by later pre-submit job when searching for a healthy and complete 
+    code coverage profile in the post-submit jobs
+    - Successfully generated coverage profile may be used as the basis of comparison for coverage change, in pre-submit workflow
+2. Read, filter, and summarizes data from coverage profile and store per-file coverage data
+    - filter based on git attribute to ignore files with the following git attributes
+      - linguist-generated
+      - coverage-excluded
+      
+      An example of how these git attribute is used can be found [here](https://github.com/knative/serving/blob/master/.gitattributes)
+    - Stores in the XML format, that is used by TestGrid, and dump it in artifacts directory
+      - The XML should be a valid JUnit XML file. See 
+  [JUnit XML format](https://www.ibm.com/support/knowledgecenter/en/SSQ2R2_14.1.0/com.ibm.rsar.analysis.codereview.cobol.doc/topics/cac_useresults_junit.html)
+      - For each file that has a coverage level lower than the threshold, the corresponding entry in the XML should have a \<failure\> tag
+
+## Pre-submit workflow
+Runs code coverage tool to report coverage change from a PR
+
+1. Generate coverage profile in artifacts directory
+2. Read, filter, and summarizes data from coverage profile and store per-file coverage data (Same as the corresponding step in post-submit)
+3. Calculate coverage changes. Compare the coverage file generated in this cycle against the most
+ recent successful post-submit build. Coverage file for post-submit commits were generated in 
+ post-submit workflow and stored in gcs bucket
+4. Use PR data from github, git-attributes, as well as coverage change data calculated above, to 
+produce a list of files that we care about in the line-by-line coverage report. produce line by 
+line coverage html and add link to covbot report. Note that covbot is the robot github account 
+used to report code coverage change results. See Covbot section for more details.
+5. If any file in this commit has a coverage change, let covbot post presubmit coverage on github, under that conversation of the PR. 
+  - The covbot comment should have the following information on each file with a coverage change
+    - file name
+    - old coverage (coverage before any change in the PR)
+    - new coverage (coverage after applied all changes in the PR)
+    - change the coverage
+  - After each new posting, any previous posting by covbot should be removed
+
+## Locally running presubmit and post-submit workflows
+Both workflows may be triggered locally in command line, as long as all the required flags are 
+supplied correctly. In addition, the following env var needs to be set:
+- JOB_TYPE (one of 'presubmit', 'postsubmit', 'local-presubmit')
+
+use 'local-presubmit' will run the presubmit workflow without posting result on github PR
+  
+## Covbot
```

Comment:
Not a huge fan of this approach -- do we have a way to determine which jobs should do something like this? Do we want every job to format their results and post them? If spyglass can have a view that exposes the job's results nicely, why is this useful?

Reply:
I don't quite understand your comment here. what do you mean by "this"? What is spyglass?

Reply:
@stevekuznetsov

Reply:
There are several reason I want to post it as comment instead of in spyglass. 

- Unlike many other jobs, of which users only care about the pass/fail status, the key information presented to the users are the code coverage change. If we put it in spyglass, users won't pay attention to it unless it fails. Plus it takes extra effort to click on the details and find the information.
- As described in pre-submit workflow, the covbot removes any previous comment, so it won't be spammy

Reply:
> Unlike many other jobs, of which users only care about the pass/fail status

I think every developer is interested in why a job fails when it does. I don't think this job is special. Furthermore, we know from experience that jobs that do not fail or jobs that are not required for merge are ignored. If you do not enforce that coverage must not decrease due to a PR, it will decrease until it hits the threshold.

> If we put it in spyglass, users won't pay attention to it unless it fails.

I would argue this is true regardless of where the output is. If the job is not required for merge, developers will simply ignore it if it is not failing.

> Plus it takes extra effort to click on the details and find the information.

This is true for every single job -- that's why I ask in my original comment if we expect every single job to start posting comments.

> As described in pre-submit workflow, the covbot removes any previous comment, so it won't be spammy

This will add one comment to every single PR that it runs on, and will update every single time the job runs. That will spam e-mails, people are already annoyed at the test failure grid.

My biggest reservations are:

 - from our experience, non-blocking jobs get ignored
 - from our experience, more comments on PRs is an anti-pattern
 - we are attempting to make Spyglass as nice as possible to solve this problem for _all_ jobs, why is this one special?

/cc @spiffxp @fejta @cjwagner would love to hear your opinions

Reply:
@jessiezcc @adrcunha @srinivashegde86 Steve is suggesting using Spyglass instead of Covbot and leave no comments on PR at all. This is against our initial consensus and design - what do you think? I tried my best to explain why we want to do it this way. I think I need to loop you all in to the discussion

Reply:
I think the key here is whether a posting on coverage change provides value, when the job succeeded. Knative repos, which has been using this code coverage tool for more than a quarter, could be a good example. Repo owners & managers do want to have the job post the code coverage change on all PRs (that has coverage change), so that developers can be aware of coverage changes and know exact impact on coverage of his commits. So far, no developer has questioned this approach and report the posting or email to be spammy. Furthermore, code coverage continuous to improve at good pace, even though blocking is not enforced - that is evidence for the effectiveness of the postings.

If we really want to use Spyglass instead, we can make the following future changes:
 - produce coverage differences in a format that is usable by Spyglass
 - make the posting optional and let the repo owner decide whether to enable it

@stevekuznetsov Please let me know if you are good with this approach

Reply:
If you have enough organizational momentum to get developers to care when the job is not failing and not blocking, then kudos to knative :)

> An additional point why we cannot go all in with Spyglass - it will to some extent defeat the effort of making code coverage tool an independent tool. It will make it reliant on Spyglass for displaying results, and thus make it harder to onboard it to repos that don't use Spyglass.

Not true -- the point of Spyglass was to let job authors integrate however they want -- place data in GCS and tell Spyglass how to render that into a view in a separate manner. You can achieve both comment reporting and Spyglass views without coupling them together.

If your org is OK with the comments, seems fine to me for now.

Reply:
One other issue @cjwagner do we care for security's sake to not mount the `covbot` token in presubmits?

Reply:
It is definitely a security concern to expose a bot token to PR code. It would be better to have a separate reporting mechanism to avoid running PR code in a job with access to the bot token. 
I pointed this out during the design phase, but writing a single job is easier than writing a job and a reporting agent so it was decided that limiting the permissions of the bot token to only be able to post comments was sufficient. This limits any malicious PRs to only being able to post comments which the malicious user's account must already have permission to do.
The main risk would be the impersonation of the @covbot account, not the ability to post comments.

If we want to split the logic into two components to improve security we can probably open an issue and do that in a follow up?

Reply:
Seems reasonable, but a lot of work to reimplement reporting instead of using what we already have. Let's keep this for the original work here and keep other options open for later.